src/encoder/mod.rs

```diff
@@ -0,0 +1,343 @@
+use std::io::{self, Write, Seek};
+use std::collections::{BTreeMap, HashMap};
+use byteorder::{ByteOrder, WriteBytesExt, NativeEndian};
+
+use crate::decoder::ifd;
+use crate::error::TiffResult;
+
+mod writer;
+use self::writer::*;
+
+pub struct Rational {
+    pub n: u32,
+    pub d: u32,
+}
+
+pub trait TiffValue {
+    const BYTE_LEN: u32;
+    const FIELD_TYPE: u16;
+    fn count(&self) -> u32;
+    fn bytes(&self) -> u32 {
+        self.count() * Self::BYTE_LEN
+    }
+    fn write<W: Write>(&self, writer: &mut TiffWriter<W>) -> TiffResult<()>;
+}
+
+impl<T: TiffValue> TiffValue for &[T] {
+    const BYTE_LEN: u32 = T::BYTE_LEN;
+    const FIELD_TYPE: u16 = T::FIELD_TYPE;
+
+    fn count(&self) -> u32 {
+        self.iter().map(|x| x.count()).sum()
+    }
+
+    fn write<W: Write>(&self, writer: &mut TiffWriter<W>) -> TiffResult<()> {
+        for x in *self {
+            x.write(writer)?;
+        }
+        Ok(())
+    }
+}
+
+impl TiffValue for u8 {
+    const BYTE_LEN: u32 = 1;
+    const FIELD_TYPE: u16 = 1;
+
+    fn count(&self) -> u32 {
+        1
+    }
+
+    fn write<W: Write>(&self, writer: &mut TiffWriter<W>) -> TiffResult<()> {
+        writer.write_u8(*self)?;
+        Ok(())
+    }
+}
+
+impl TiffValue for u16 {
+    const BYTE_LEN: u32 = 2;
+    const FIELD_TYPE: u16 = 3;
+
+    fn count(&self) -> u32 {
+        1
+    }
+
+    fn write<W: Write>(&self, writer: &mut TiffWriter<W>) -> TiffResult<()> {
+        writer.write_u16(*self)?;
+        Ok(())
+    }
+}
+
+impl TiffValue for u32 {
+    const BYTE_LEN: u32 = 4;
+    const FIELD_TYPE: u16 = 4;
+
+    fn count(&self) -> u32 {
+        1
+    }
+
+    fn write<W: Write>(&self, writer: &mut TiffWriter<W>) -> TiffResult<()> {
+        writer.write_u32(*self)?;
+        Ok(())
+    }
+}
+
+impl TiffValue for Rational {
+    const BYTE_LEN: u32 = 8;
+    const FIELD_TYPE: u16 = 5;
+
+    fn count(&self) -> u32 {
+        1
+    }
+
+    fn write<W: Write>(&self, writer: &mut TiffWriter<W>) -> TiffResult<()> {
+        writer.write_u32(self.n)?;
+        writer.write_u32(self.d)?;
+        Ok(())
+    }
+}
+
+impl TiffValue for str {
+    const BYTE_LEN: u32 = 1;
+    const FIELD_TYPE: u16 = 2;
+
+    fn count(&self) -> u32 {
+        self.len() as u32 + 1
+    }
+
+    fn write<W: Write>(&self, writer: &mut TiffWriter<W>) -> TiffResult<()> {
+        for x in self.chars() {
+            if x.is_ascii() {
```

Comment:
Would it make more sense to return an error if the string is not ASCII instead of silently skipping characters ? Doing so and checking the whole string at once, the function could be rewritten as something like:

```
if self.is_ascii() {
    writer.write_bytes(self.as_bytes())?;
    Ok(writer.write_u8(0)?)
}
else {
    Err(...)
}
```